src/bin/pg_autoctl/pgctl.c

```diff
@@ -1328,44 +1325,282 @@ pg_setup_standby_mode(uint32_t pg_control_version,
 {
 	char recoveryConfPath[MAXPGPATH];
-	PQExpBuffer content = NULL;
 
-	log_trace("pg_write_recovery_conf");
+	/* prepare storage areas for parameters */
+	char primaryConnInfo[MAXCONNINFO] = { 0 };
+	char primarySlotName[MAXCONNINFO] = { 0 };
+	char targetLSN[PG_LSN_MAXLENGTH] = { 0 };
 
-	/* build the contents of recovery.conf */
-	content = createPQExpBuffer();
-	appendPQExpBuffer(content, "standby_mode = 'on'");
-	appendPQExpBuffer(content, "\nprimary_conninfo = %s", primaryConnInfo);
-	appendPQExpBuffer(content, "\nprimary_slot_name = '%s'", replicationSlotName);
-	appendPQExpBuffer(content, "\nrecovery_target_timeline = 'latest'");
-	appendPQExpBuffer(content, "\n");
+	GUC recoverySettingsStandby[] = {
+		{ "standby_mode", "'on'" },
+		{ "primary_conninfo", (char *) primaryConnInfo },
+		{ "primary_slot_name", (char *) primarySlotName },
+		{ "recovery_target_timeline", "'latest'" },
+		{ NULL, NULL }
+	};
 
-	/* memory allocation could have failed while building string */
-	if (PQExpBufferBroken(content))
+	GUC recoverySettingsTargetLSN[] = {
+		{ "standby_mode", "'on'" },
+		{ "primary_conninfo", (char *) primaryConnInfo },
+		{ "primary_slot_name", (char *) primarySlotName },
+		{ "recovery_target_timeline", "'latest'" },
+		{ "recovery_target_lsn", (char *) targetLSN },
+		{ "recovery_target_inclusive", "'true'" },
+		{ "recovery_target_action", "'pause'" },
+		{ NULL, NULL }
+	};
+
+	GUC *recoverySettings =
+		IS_EMPTY_STRING_BUFFER(replicationSource->targetLSN)
+		? recoverySettingsStandby
+		: recoverySettingsTargetLSN;
+
+	join_path_components(recoveryConfPath, pgdata, "recovery.conf");
+
+	log_info("Writing recovery configuration to \"%s\"", recoveryConfPath);
+
+	if (!prepare_recovery_settings(pgdata,
+								   replicationSource,
+								   recoverySettings,
+								   primaryConnInfo,
+								   primarySlotName,
+								   targetLSN))
 	{
-		log_error("Failed to allocate memory");
-		destroyPQExpBuffer(content);
+		/* errors have already been logged */
 		return false;
 	}
 
-	join_path_components(recoveryConfPath, pgdata, "recovery.conf");
+	return ensure_default_settings_file_exists(recoveryConfPath,
+											   recoverySettings,
+											   NULL);
+}
 
-	log_info("Writing recovery configuration to \"%s\"", recoveryConfPath);
-	log_debug("%s:\n%s", recoveryConfPath, content->data);
 
-	if (!write_file(content->data, content->len, recoveryConfPath))
+/*
+ * pg_write_standby_signal writes the ${PGDATA}/standby.signal file that is in
+ * use starting with Postgres 12 for starting a standby server. The file only
+ * needs to exists, and the setup is to be found in the main Postgres
+ * configuration file.
+ */
+static bool
+pg_write_standby_signal(const char *pgdata,
+						ReplicationSource *replicationSource)
+{
+	char standbyConfigFilePath[MAXPGPATH] = { 0 };
+	char signalFilePath[MAXPGPATH] = { 0 };
+	char configFilePath[MAXPGPATH] = { 0 };
+
+	/* prepare storage areas for parameters */
```

Comment:
This technique looks a little weird at first, yes. The idea is that we don't allocate memory on the heap: we avoid malloc so that we don't forget to free. So having those string buffers in the stack, we then pass pointers to those pre-allocated strings to a shared function that fills-in the buffers for us.

The other reason why we do it that way is that we don't want to loop on the GUC list and string compare the name of the GUC to know which string buffer to fill-in when we have proper pointers to the memory areas to take care of.